plugins/logging/pkg/backend/metadata.go

```diff
@@ -0,0 +1,227 @@
+				if resp.StatusCode != http.StatusNotFound {
+					b.Logger.With(
+						"cluster", cluster.Id,
+						zap.Error(err),
+					).Errorf("could not get metadata document for cluster")
+					return
+				}
+
+				b.Logger.With(
+					"cluster", cluster.Id,
+				).Debug("no metadata found for cluster")
+
+				if err := b.addClusterMetadata(ctx, cluster.Reference()); err != nil {
+					b.Logger.With(
+						"cluster", cluster.Id,
+						zap.Error(err),
+					).Errorf("")
+				}
+			}
+
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+
+	b.Logger.Infof("reconciled logging cluster metadata")
+
+	return nil
+}
+
+func (b *LoggingBackend) addClusterMetadata(ctx context.Context, ref *opnicorev1.Reference) error {
+	cluster, err := b.MgmtClient.GetCluster(ctx, ref)
+	if err != nil {
+		return fmt.Errorf("could not get data for cluster '%s': %w", ref.Id, err)
+	}
+
+	clusterName, found := cluster.Metadata.Labels[opnicorev1.NameLabel]
+	if !found {
+		clusterName = cluster.Id
+	}
+
+	resp, err := b.OpensearchManager.Indices.AddDocument(ctx, multiclusterrolebinding.ClusterMetadataIndexName, cluster.Id, opensearchutil.NewJSONReader(map[string]string{
+		"id":   cluster.Id,
+		"name": clusterName,
+	}))
+
+	if err != nil {
+		return fmt.Errorf("could not add cluster '%s' to metadata index: %w", clusterName, err)
+	}
+
+	if resp.IsError() {
+		errMsg, err := io.ReadAll(resp.Body)
+		resp.Body.Close()
+
+		if err != nil {
+			return fmt.Errorf("could not add cluster '%s' to metadata index: %w", clusterName, err)
+		}
+
+		return fmt.Errorf("could not add cluster '%s' to metadata index: %s", clusterName, errMsg)
+	}
+
+	resp.Body.Close()
+
+	b.Logger.With(
+		"cluster", ref.Id,
+	).Infof("added cluster to metadata index")
+
+	return nil
+}
+
+func (b *LoggingBackend) updateClusterMetadata(ctx context.Context, event *managementv1.WatchEvent) error {
+	clusterId := event.Cluster.Id
+
+	newName, oldName := event.Cluster.Metadata.Labels[opnicorev1.NameLabel], event.PreviousCluster.Metadata.Labels[opnicorev1.NameLabel]
+	if newName == oldName {
+		b.Logger.With(
+			"oldName", oldName,
+			"newName", newName,
+		).Debug("cluster was not renamed")
+		return nil
+	}
+
+	b.Logger.With(
+		"oldName", oldName,
+		"newName", newName,
+	).Debug("cluster was renamed")
+
+	if b.OpensearchManager.Client == nil {
+		b.Logger.With(
+			"cluster", clusterId,
+		).Warnf("plugin has nil opensearch client, doing nothing")
+		return nil
+	}
+
+	resp, err := b.OpensearchManager.Indices.UpdateDocument(ctx, multiclusterrolebinding.ClusterMetadataIndexName, clusterId, opensearchutil.NewJSONReader(
+		types.MetadataUpdate{
+			Document: types.ClusterMetadataUpdate{
+				Name: newName,
+			},
+		},
+	))
+
+	if err != nil {
+		b.Logger.With(zap.Error(err)).Errorf("failed to update cluster in metadata index")
+		return nil
+	}
+
+	if resp.IsError() {
+		errMsg, err := io.ReadAll(resp.Body)
+
+		if err != nil {
+			b.Logger.With(zap.Error(err)).Errorf("failed to read response body")
+			return nil
```

Comment:
potential resource leak in this path : resp.Body is never closed
for the functions in this file, consider using the pattern:
```go
resp, err := // ...
if err != nil {
  // ...
}
defer resp.Body.Close
```

Reply:
I call `resp.Body.Close()` down below, but deferring is definitely more idiomatic

Reply:
there is a `return nil` in that particular code path that results in the resp.Body.Close below not being closed